### PR TITLE
GH-2211 add more latches and loop to reduce chance of flaky failure

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #2211 

Briefly describe the changes proposed in this PR:

---- 
PR Author Checklist: 

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

File: core/model/src/test/java/org/eclipse/rdf4j/model/DynamicModelConcurrentModificationAndUpgradeTest.java [hunk 1]
```diff
@@ -75,20 +91,23 @@ public boolean contains(Object o) {
 							@Override
 							public boolean hasNext() {
 								try {
-									countDownLatch.await();
+									countDownLatch1.countDown();
+									countDownLatch2.await();
```
Comment (hmottestad) [inline core/model/src/test/java/org/eclipse/rdf4j/model/DynamicModelConcurrentModificationAndUpgradeTest.java:95]: This latching here should help reduce the chance of this one thread getting executed first.

File: core/model/src/test/java/org/eclipse/rdf4j/model/DynamicModelConcurrentModificationAndUpgradeTest.java [hunk 2]
```diff
@@ -145,8 +164,13 @@ public void clear() {
...
 
 		Runnable upgrade = () -> {
+			try {
+				countDownLatch1.await();
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
 			model.filter(null, RDF.TYPE, null);
-			countDownLatch.countDown();
+			countDownLatch2.countDown();
```
Comment (hmottestad) [inline core/model/src/test/java/org/eclipse/rdf4j/model/DynamicModelConcurrentModificationAndUpgradeTest.java:173]: and here is the count down for that latch

**Comment by hmottestad** [comment]
@jeenbroekstra I don't really want that Set to be volatile, for this test to be correct it really should be, but I think running the test several times and adding yield will sufficiently increase the odds of reading the correct variable at some point during the test. I've also fixed a bigger issue with the test. If run on a machine with only 1 core then the first thread would just finish executing and then the second would be scheduled. I've added a second latch to force switching from one thread to the other.

**Comment by jeenbroekstra** [comment]
If this gets things stable again I'm fine with it, but isn't this just rigging the test rather than fixing the underlying issue? Or are you saying that it's the test itself that's flawed?

**Comment by hmottestad** [comment]
yeah, test is to check for the dynamicmodel best-effort concurrent modification detection

**Pull request merged**